File: 0813-all-paths-from-source-to-target/0813-all-paths-from-source-to-target.cpp
class Solution {
public:
    vector<vector<int>> pathes{};
    
    void find_pathes(vector<vector<int>>& graph, vector<int> path = {}, int i = 0){
        path.push_back(i);

        if(i == graph.size() - 1){
            pathes.push_back(path);
            return;
        }

        for(int j = 0; j < graph[i].size(); ++j){
            find_pathes(graph, path, graph[i][j]);
        }
    }

    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        find_pathes(graph);

        return pathes;
    }
};

File: 0813-all-paths-from-source-to-target/0813-all-paths-from-source-to-target.erl
-spec all_paths_source_target(Graph :: [[integer()]]) -> [[integer()]].
all_paths_source_target(Graph) ->
    dfs(Graph, length(Graph), 1, []).

dfs(_    , N, N, Path) -> [Path ++ [N - 1]];
dfs(Graph, N, I, Path) ->
    NextPath = Path ++ [I - 1],
    lists:foldl(
        fun(X, Res) ->
            Res ++ dfs(Graph, N, X + 1, NextPath)
        end, [], lists:nth(I, Graph)
    ).
    